README.md
# Faxfrontend

## 1. Introduction


Faxfrontend is a GUI interface to the HylaFAX (tm) client programs. It includes a modified version of Sebastiano Vigna's fax4CUPS to invoke it when a file is printed to a CUPS printer. With the fax4CUPS interface installed, the user prints to the fax printer, and the GUI front end is invoked to get the fax number, etc., from the user. The program can also be invoked in a standalone mode to queue a file for faxing.

A rolodex for commonly used fax numbers is provided, and a query to the HylaFAX server is displayed after the fax is queued for transmission.

When the files have been handed off to HylaFAX to send, the program displays the fax queue status by repeatedly executing 'faxstat', and capturing the output for display:

## 2. Requirements

1. Qt libraries, moc, and uic utilities ( http://www.trolltech.com )
2. `sudo` - execute a command as another user ( http://www.sudo.ws/sudo/sudo.html )
3. CUPS - Common Unix Printing System ( http://www.cups.org )

## 3. Installation

### 3.2. Building the distribution

#### 3.2.1. Qt development files

Faxfrontend uses the Qt GUI tool kit, version 3, and the QTDIR environment variable must be properly set before starting the build. It should point to the directory containing the Qt header files and libraries, so a typical setting is:

    export QTDIR=/usr/lib/qt3

and so `/usr/lib/qt3/include` would hold the headers, `/usr/lib/qt3/bin` contains the `moc` and `uic` executables. Alternatively, use `configure –with-qtdir=/usr/lib/qt3` in the next step. The Ubuntu distribution uses /usr/share/qt3, so `configure –with-qtdir=/usr/share/qt3` should work.

#### 3.2.2. Build and install

    export QTDIR=<qt-distrib-dir>
    configure
    make
    make install
    
### 3.3. System setup

#### 3.3.1 `sudo` configuration

The CUPS system must be able run the `sudo` command in a backend script. The account that the CUPS daemon `cupsd` runs with must be enabled to run `sudo` command without passwords. The account used by CUPS to run external programs such as backend scripts is controlled by the `User` directive in `etc/cups/cupsd.conf`. For example, if the CUPS daemon uses the 'lp' account, there must be a line in `/etc/sudoers` to enable `sudo` command usage:

    lp ALL=(ALL) NOPASSWD: ALL

Use the `visudo` command to edit the `/etc/sudoers` file.

#### 3.3.2 CUPS configuration

Restart the CUPS server, and define a fax printer for CUPS to use the hylafax.ppd and hylafax backend script. Use the CUPS WWW tool to create the printer ( http://localhost:631 ). If the files have been correctly installed, the screens should look like the following screen shots:

## 4. Standalone Usage

Invoke from the command line, either with or without a filename for faxing:

    faxfrontend [file to fax]

If no file is given, the user has the option to send a cover page only, using the comments section for the fax content.

## 5. File Manifest

Executables:

    /usr/local/bin/faxfrontend
    /usr/local/bin/messagebox

CUPS configuration files:

    /usr/lib/cups/backend/hylafax
    /usr/share/cups/model/hylafax.ppd

Defaults file:

    ~/.faxfrontendrc

Rolodex file:

    ~/.faxrolodex

## 6. Troubleshooting

The CUPS error log (usually `/var/log/cups/error_log`) can contain useful information.  Logging is controlled by the `/etc/cupsd.conf` file.  The `faxfrontend` program will log the faxcover and sendfax commands it generates to the system log (level `daemon.info`) when executed with the `-d` option.  Change the line in `/usr/lib/cups/backend/hylafax that executes the `faxfrontend` program to

    m=`sudo -H -u $USER $FAXGUI -d $FAXNAME`

to enable debug logging.

When the backend script `usr/lib/cups/backend/hylafax` is executed at the command line with no arguments, it should report standard CUPS backend `discovery` information, and exit with status 0:

    # /usr/lib/cups/backend/hylafax 

direct hylafax:`/local "Unknown" "Local HylaFAX server"`

## 7. Acknowledgments

This is a Qt knock off of Andy Moskoff's `tkhylafax` ( ftp://ftp.hylafax.org/contrib/tkhylafax ). I really wanted to use the fax4CUPS interface, but wasn't adept enough with Tcl/Tk to make modifications to `tkhylafax` to accept a filename as an argument. Sebastiano Vigna did the original work for fax4CUPS ( http://gongolo.usr.dsi.unimi.it/~vigna/fax4CUPS ), which is really the best paradigm for sending a fax. Geoffrey Wossum supplied configuration tools for Qt with autoconf ( http://sourceforge.net/projects/autoqt ).

Glenn Burkhardt
January, 2004
Westford, Massachusetts

gbburkhardt@verizon.net
ftp://ftp.hylafax.org/contrib/faxfrontend


ChangeLog
1.1; 22 Mar 2007 Explicitly pass DISPLAY variable in backend script 
	         in showerror() function to avoid 'Default env_reset' in
		 /etc/sudoers.  Add check that 'sudo' command can be
		 executed, since some distributions are running CUPS under
		 'lp' user, instead of 'root' user.  Update documentation
		 to note that account that CUPS uses to run must be
		 enabled to run 'sudo' commands with NOPASSWD.  Correct
		 file permission problem (when CUPS runs with a
		 non-superuser account, it can't change the owner of the
		 temporary file used to access the fax by the GUI).
		 Add check for correct DISPLAY variable
		 
1.0; 28 Feb 2006 Tweak backend script to avoid problem with /etc/sudoers
		'Default env_reset' setting (thanks to Karl Schmidt)

0.7; 6 Nov 2004 Copy the file passed to CUPS backend as arg 6 to a temporary
		file so it is readable by all users; security levels on some
		systems prevent users from reading files from the CUPS spool
		directory, where the file lives.

0.6;24 Aug 2004 Eliminated use of Qt Platinum style; not available with 
	        version of Qt shipped with Mandrake 10.0

0.5: 5 Apr 2004 Fixed bug:  backend script wouldn't ignore comments
	        in hyla.conf file; line with "#Host: foobar.com" wouldn't
		be ignored.

0.4:12 Feb 2004 Fixed bugs:  couldn't have single quotes in remarks, etc.
		Internal cover page wouldn't send if 'no cover page'
		was checked.  Internal cover page generation didn't use
		page size.  Correct need to exit twice from status display.
		Additional error checking.  Capture stderr from faxstat
		for display.  Fork status display so print queue isn't
		hung while waiting for user to quit status display.

0.3:25 Jan 2004 Check port to Solaris, improve check for MOC, UIC in
	      	acsite.m4, minor fix for Qt2, removed moc files from
	      	distribution (should be re-built at each site); 
		corrected backend script to limit prints to stdout
		so CUPS gets correct device info.  Corrected documentation
		to reflect what happens with correct backend script.

0.2:19 Jan 2004 Improved documentation
0.1:19 Jan 2004 Initial version.


ffe/main.cc
/***************************************************************************
$Id: main.cc,v 1.5 2004/08/24 20:28:23 glenn Exp $

                          main.cc - Fax Front End for Hylafax
                             -------------------
    begin                : Thu Jan 8, 2004
    copyright            : (C) 2004 by Glenn Burkhardt
    email                : gbburkhardt@verizon.net

    Usage:  faxfrontend [-debug] [file to fax]

            If invoked with no arguments, user is allowed to enter
            fax file or send cover page only.

 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <syslog.h>
#include <sys/wait.h>
#include <sys/ioctl.h>
#include <sys/fcntl.h>
#include <pwd.h>
#include <qapplication.h>
#include <qfileinfo.h>
#include <qfont.h>
#include <qstring.h>
#include <qtextcodec.h>
#include <qtranslator.h>
#include <qcheckbox.h>
#include <qcombobox.h>
#include <qmessagebox.h>
#include <qbuttongroup.h>
#include <qmultilineedit.h>
#include <qspinbox.h>
#include <qlineedit.h>
#include "FaxFrontEnd.h"
#include "faxstat.h"

/* Scan the string for single quotes, escape them, and return a
   double quoted string suitable for passing the string as a shell argument.
*/
   
static QString quoted_string(QString str)
{
    for (int i=0; i<str.length(); i++) {
	if ((str[i] == '\\') || (str[i] == '`')
	    || (str[i] == '$') || (str[i] == '"')) {

	    str.insert(i, '\\');
	    i += 2;
	}
    }

    str.prepend('"');
    str.append('"');

    return str;
}

/* Check that the file exists, is a regular file, and we can read it.
 */
static bool file_OK(const QString& fn)
{
    QFileInfo info(fn);

    if (!info.exists()) {
	QMessageBox::critical(0, QObject::tr("File Access Error"),
			QObject::tr("Fax File %1 doesn't exist.").arg(fn));
	return false;
    }

    if (info.isDir()) {
	QMessageBox::critical(0, QObject::tr("File Access Error"),
			QObject::tr("%1 is a directory").arg(fn));
	return false;
    }

    if (!info.isReadable()) {
	QMessageBox::critical(0, QObject::tr("File Access Error"),
			QObject::tr("%1 is not readable").arg(fn));
	return false;
    }

    return true;
}

int main(int ac, char *av[])
{
    char tmpname[] = "/tmp/FaxCoverXXXXXX";
    char *user, *faxfile;
    bool delete_tmpname=false, debug=false;
    int i, status;
    struct passwd *pw;

    // allocation is from heap, so child faxstat process can tag along
    // on the X connection.
    QApplication* a = new QApplication(ac, av);

    // set the location where your .qm files are in load() below as the 
    // last parameter instead of "."
    // for development, use "/" to use the english original as
    // .qm files are stored in the base project directory.
    QTranslator tor( 0 );
    QFileInfo fi(av[0]);
    tor.load( QString("faxfrontend.")
	      + QTextCodec::locale(), fi.dirPath(true) );
    a->installTranslator( &tor );

    /* Check for arguments: -d for debug, file to fax */
    faxfile = 0;
    for (int i=1; i<ac; i++) {
	if (av[i][0] == '-' && av[i][1] == 'd') {
	    debug = true;
	    openlog("faxfrontend", 0, LOG_DAEMON);
	} else {
	    faxfile = av[i];
	    if (!file_OK(faxfile)) faxfile = 0;
	}
    }

    FaxFrontEnd faxfrontend(faxfile);

    for (;;) {
	status = faxfrontend.exec();
	if (status == QDialog::Rejected) return 0;
	if (faxfile) break;	// already checked file status

	if (faxfrontend.fileLineEdit->text().isEmpty()) break;
	if (file_OK(faxfrontend.fileLineEdit->text())) break;
    }
    
    // send the fax

    QString cmd("sendfax");
  	
    /* Determine send mode:
       - have fax file as argument, or given in file line edit,
       use cover page option
       - No file given, offer to build cover page only and send.
       
       Must be careful - arguments to sendfax are order dependent.
    */
    
    if (faxfile || !faxfrontend.fileLineEdit->text().isEmpty()) {
	// Have a file to fax

	if (faxfrontend.noCoverPage->isChecked()) {
	    cmd += " -n";
	} else {
	    if (!faxfrontend.companyBox->text().isEmpty())
		cmd += " -x "
		    + quoted_string(faxfrontend.companyBox->text()
				    .stripWhiteSpace());
	    if (!faxfrontend.locationBox->text().isEmpty())
		cmd += " -y "
		    + quoted_string(faxfrontend.locationBox->text()
				    .stripWhiteSpace());
	    if (!faxfrontend.regardingBox->text().isEmpty())
		cmd += " -r "
		    + quoted_string(faxfrontend.regardingBox->text()
				    .stripWhiteSpace());
	    if (!faxfrontend.comments->text().isEmpty())
		cmd += " -c "
		    + quoted_string(faxfrontend.comments->text()
				    .stripWhiteSpace());
	}
    } else
	cmd += " -n";	// no file to fax; assume cover page generation
			// option (below), specify sendfax no cover page
    
    if (faxfrontend.notifyComplete->isChecked())
	cmd += " -D";
    else
	cmd += " -N";
    if (faxfrontend.notifyRetry->isChecked()) cmd += " -R";
    if (faxfrontend.resButtonGroup->selected()
	== (QButton*)faxfrontend.lowResButton) cmd += " -l";
    
    // transmit delay value
    if (faxfrontend.xmitSpinBox->value() > 0) {
	cmd += QString(" -a 'now + %1 minutes'")
	    .arg(faxfrontend.xmitSpinBox->value());
    }
    if (faxfrontend.retrySpinBox->value() > 1) {
	QString retry;
	retry.sprintf(" -t %d", faxfrontend.retrySpinBox->value());
	cmd += retry;
    }
    if (faxfrontend.pageSizeComboBox->currentItem())
	cmd += " -s a4";
    else
	cmd += " -s na-let";
    
    /* Destination is 'user@fax-number'; strip '@' from user name
       just to be sure.
    */
    QString qs(faxfrontend.recipientBox->text().stripWhiteSpace());
    i = 0;
    while ((i = qs.find('@', i)) >= 0) qs[i++] = '_';

    qs += "@" + faxfrontend.faxNumberBox->text().stripWhiteSpace();
    
    cmd += " -d " + quoted_string(qs);
    
    cmd += " ";	// to precede filename
    
    /* If no arguments were given, assume test mode */
    if (faxfile)
	cmd += faxfile;	// fax file name
    else if (!faxfrontend.fileLineEdit->text().isEmpty())
	cmd += faxfrontend.fileLineEdit->text();
    else {
	/* Send cover page only */
	if (QMessageBox::warning(0, QObject::tr("Warning"),
				 QObject::tr("No file to fax specified.\n"
					     "Send Cover Page only?"),
				 QMessageBox::Yes|QMessageBox::Default,
				 QMessageBox::No) == QMessageBox::No) {
	    return 0;
	}
	
	mktemp(tmpname);
	QString fcmd("faxcover");
	
	if (faxfrontend.pageSizeComboBox->currentItem())
	    fcmd += " -s a4";
	else
	    fcmd += " -s na-let";
	
	fcmd += " -t " + quoted_string(faxfrontend.recipientBox->text()
				       .stripWhiteSpace());
	
	if (!faxfrontend.companyBox->text().isEmpty())
	    fcmd += " -x " + quoted_string(faxfrontend.companyBox->text()
					   .stripWhiteSpace());
	
	if (!faxfrontend.locationBox->text().isEmpty())
	    fcmd += " -l " + quoted_string(faxfrontend.locationBox->text()
					   .stripWhiteSpace());
	
	if (!faxfrontend.regardingBox->text().isEmpty())
	    fcmd += " -r " + quoted_string(faxfrontend.regardingBox->text()
					   .stripWhiteSpace());
	
	if (!faxfrontend.comments->text().isEmpty())
	    fcmd += " -c " + quoted_string(faxfrontend.comments->text()
					   .stripWhiteSpace());
	fcmd += " -p 0";
	
	/* Get 'real user' name for "from" address */
	user = getenv("USER");
	if (!user) user = "root";
	pw = getpwnam(user);
	if (!pw) 
	    fcmd += " -f HylaFAX";
	else {		
	    fcmd += " -f '";
	    fcmd += pw->pw_gecos;
	    fcmd += "'";
	}
	
	fcmd += " -n '"
	    + faxfrontend.faxNumberBox->text().stripWhiteSpace() + "'";
	
	fcmd += " > ";
	fcmd += tmpname;
	
	if (debug) {
	    syslog(LOG_INFO, fcmd.latin1());
	}
	
	status = system(fcmd.latin1());	// generate cover page
	status = WEXITSTATUS(status);
	if (status) {
	    QMessageBox::critical(0, QObject::tr("Error"),
		QObject::tr("Can't create faxcover\ncommand: %1").arg(fcmd));
	    unlink(tmpname);
	    return -1;
	}
	
	delete_tmpname = true;
	
	cmd += tmpname;	// file to fax is cover page we generated
    }
    
    if (debug) syslog(LOG_INFO, cmd.latin1());
	
    status = system(cmd.latin1());	// send the fax
    status = WEXITSTATUS(status);
    if (status) {
	QMessageBox::critical(0, QObject::tr("Error"),
			      QObject::tr("Can't send fax\ncommand: %1")
			      .arg(cmd));
    }
    
    if (delete_tmpname) unlink(tmpname);

    // Now poll status from Hylafax. Fork, and let parent return
    // completion status to CUPS so the print queue isn't held up
    // until the user closes the status window.
    if (fork()) return status;

    // detach from parent

    ::close(0);
    ::close(1);
    ::close(2);
    int fd = open("/dev/null", O_RDWR);
    dup2(fd, fileno(stdout));
    dup2(fd, fileno(stderr));
    fd = open("/dev/tty", O_RDWR);
    if (fd > 0) {
	ioctl(fd, TIOCNOTTY, 0);
	::close(fd);
    }

    FaxStat fs;
    fs.exec();
    return 0;

}
